File: apps/web/lib/sentry.ts
/**
 * spec 006 — Sentry 初始化(浏览器)
 *
 * 云版优先(sentry.io 免费层 5K 错误/月)
 * 若需自托管,改 NEXT_PUBLIC_SENTRY_DSN + 自部署 Sentry 服务
 */
export const SENTRY_DSN = process.env.NEXT_PUBLIC_SENTRY_DSN || '';
export const SENTRY_ENV = process.env.NEXT_PUBLIC_SENTRY_ENV || 'production';

export async function initSentryBrowser(): Promise<void> {
  if (typeof window === 'undefined' || !SENTRY_DSN) return;
  try {
    const Sentry = await import('@sentry/nextjs');
    Sentry.init({
      dsn: SENTRY_DSN,
      tracesSampleRate: 0.1, // 生产 10%,开发 100%(由 NEXT_PUBLIC_SENTRY_ENV 切换)
      environment: SENTRY_ENV,
      release: process.env.NEXT_PUBLIC_SENTRY_RELEASE, // Git SHA 短码
      // 隐私:不在事件中存 PII
      beforeSend(event) {
        if (event.user) {
          // 仅保留 user_id_hash(由调用方注入)
          event.user = { id: event.user.id };
        }
        return event;
      },
    });
  } catch (e) {
    // 静默失败,不影响主流程
    if (import.meta.env.DEV) console.warn('Sentry init failed:', e);
  }
}

File: apps/api/app/core/sentry.py
"""spec 006 — Sentry SDK 初始化(后端)

云版优先(sentry.io 免费层 5K 错误/月)
"""
import logging
from typing import Any

log = logging.getLogger(__name__)

_initialized = False


def init_sentry() -> None:
    """启动时调用,初始化 Sentry SDK(幂等)"""
    global _initialized
    if _initialized:
        return

    import os

    dsn = os.environ.get("SENTRY_DSN", "")
    if not dsn:
        log.info("sentry.disabled", extra={"reason": "SENTRY_DSN not set"})
        return

    try:
        import sentry_sdk
        from sentry_sdk.integrations.fastapi import FastApiIntegration
        from sentry_sdk.integrations.logging import LoggingIntegration

        sentry_sdk.init(
            dsn=dsn,
            environment=os.environ.get("SENTRY_ENV", "production"),
            release=os.environ.get("SENTRY_RELEASE"),
            traces_sample_rate=0.1,
            integrations=[
                FastApiIntegration(),
                LoggingIntegration(level=logging.INFO, event_level=logging.ERROR),
            ],
            before_send=_scrub_pii,
        )
        _initialized = True
        log.info("sentry.initialized")
    except ImportError:
        log.warning("sentry.disabled", extra={"reason": "sentry-sdk not installed"})
    except Exception as e:
        log.warning("sentry.init_failed", extra={"error": str(e)})


def _scrub_pii(event: dict[str, Any], hint: Any) -> dict[str, Any] | None:
    """去除 PII(邮箱/密码/IP 原始值)"""
    if "user" in event and event["user"]:
        user = event["user"]
        # 仅保留 user_id_hash,去除 email / username / ip_address
        if "email" in user:
            del user["email"]
        if "username" in user:
            del user["username"]
        if "ip_address" in user:
            del user["ip_address"]
    return event


def capture_exception(error: Exception, **kwargs: Any) -> None:
    """便捷上报异常(失败静默)"""
    if not _initialized:
        return
    try:
        import sentry_sdk
        sentry_sdk.capture_exception(error, **kwargs)
    except Exception:
        pass